### PR TITLE
Optimize CPU RNN performance with MKL-DNN

### PR DESCRIPTION
**Pull request by mingfeima**
This PR integrate MKLDNN RNN APIs into PyTorch to optimize CPU RNN performance

1. **Features and limitations**

- Supports both training and inference.
- MKL-DNN supports 4 standard flavors of RNN: `RNN_TANH`, `RNN_RELU`, `GRU`, `LSTM`; GRU is of cuDNN flavor.
- Supports multi-layer and bidirectional RNN, `nn.LSTM` and `nn.GRU` will call MKL-DNN if MKL-DNN is available and is able to deal the input configuration. `nn.LSTMCell` and `nn.GRUCell` won't call MKL-DNN.
- MKL-DNN doesn't support `dropout`, i will enable `dropout` later on from framework side.
- Currently only ` unpacked input` (TNC layout) will call MKL-DNN and `packed input` is not supported. i will enable `packed input` later on from framework side.

2. **Performance**

I wrote [rnn-benchmark](https://github.com/mingfeima/pytorch-rnn-benchmark) to compare the performance `with`/`without` MKL-DNN.
This PR adds a switch `torch._C._set_mkldnn_enabled` to turn `on`/`off` MKL-DNN functionalities, the usage is identical to the `cudnn` equivalent `torch._C._set_cudnn_enabled`.

Below is the benchmark result on Xeon 6140 (20*2 cores @2.5GHz), approximately `1.5~2x` performance improvement.
```bash
(pytorch-mingfei) [mingfeim@mlt-skx086 pytorch-rnn-benchmark]$ ./run.sh

### training mode

### using OMP_NUM_THREADS=40
### using KMP_AFFINITY=granularity=fine,compact,1,0
### using KMP_BLOCKTIME=1

mkldnn: [T,N,I,H] = [35,32,500,500], time: 29.66 (ms), SPS: 1078.98
cpu   : [T,N,I,H] = [35,32,500,500], time: 60.81 (ms), SPS: 526.22
mkldnn: [T,N,I,H] = [35,32,800,800], time: 69.83 (ms), SPS: 458.23
cpu   : [T,N,I,H] = [35,32,800,800], time: 99.90 (ms), SPS: 320.32
mkldnn: [T,N,I,H] = [35,64,500,500], time: 48.16 (ms), SPS: 1329.01
cpu   : [T,N,I,H] = [35,64,500,500], time: 81.94 (ms), SPS: 781.08
mkldnn: [T,N,I,H] = [35,64,800,800], time: 86.35 (ms), SPS: 741.16
cpu   : [T,N,I,H] = [35,64,800,800], time: 120.99 (ms), SPS: 528.96
mkldnn: [T,N,I,H] = [50,32,500,500], time: 45.25 (ms), SPS: 707.12
cpu   : [T,N,I,H] = [50,32,500,500], time: 73.30 (ms), SPS: 436.54
mkldnn: [T,N,I,H] = [50,32,800,800], time: 72.25 (ms), SPS: 442.91
cpu   : [T,N,I,H] = [50,32,800,800], time: 145.42 (ms), SPS: 220.05
mkldnn: [T,N,I,H] = [50,64,500,500], time: 60.91 (ms), SPS: 1050.66
cpu   : [T,N,I,H] = [50,64,500,500], time: 95.31 (ms), SPS: 671.49
mkldnn: [T,N,I,H] = [50,64,800,800], time: 113.79 (ms), SPS: 562.42
cpu   : [T,N,I,H] = [50,64,800,800], time: 192.17 (ms), SPS: 333.04
```

For your production models, you can toggle the flag `torch._C._set_mkldnn_enabled` to see how MKL-DNN performs, and `export MKLDNN_VERBOSE=2` will give you more performance details on MKL-DNN. In case MKL-DNN doesn't meet your expectation, let me know. I will try to solve it inside Intel.

3. **Validation**

i wrote [pytorch-rnn-test](https://github.com/mingfeima/pytorch-rnn-test) to validate MKL-DNN RNN APIs. The test case will verify whether MKL-DNN output are `identical` to CPU output. The following are fully testes:
- `output`, `hx`, `cx` of forward path
- `grad_input`, `grad_hx`, `grad_cx`, `grad_weight_ih`, `grad_weight_hh`, `grad_ih`, `grad_hh` of backward path

4. **Known issue**

The current MKL-DNN version (`v0.17`, commit id `830a100`) on PyTorch has issue in workspace offset calculation on backward when `input_size != hidden_size` or `bidirectional=True`. So Training on these circumstances are intentionally banned at the moment. Inference is not affected, only training has such issue.

This issue has been **FIXED** on [mkl-dnn](https://github.com/intel/mkl-dnn) master already.
Once the MKL-DNN version on PyTorch is upgraded, say `v0.18`, i will release these limitations. 

5. **Future work**

Aside from `TODO`s previously mentioned in this thread, i will provide optimized fused kernels on CPU device shortly, similar to `_thnn_fused_lstm_cell` and `_thnn_fused_gru_cell`. This will be beneficial for `nn.LSTMCell` and `nn.GRUCell`, as well as cases of `nn.LSTM` and `nn.GRU` when MKL-DNN fails to handle.

### CODE REVIEW

**Comment by ezyang** [comment]
Build errors are real.

**Comment by ezyang** [comment]
```
Jan 15 01:11:14 make[1]: *** Waiting for unfinished jobs....
Jan 15 01:11:14 /var/lib/jenkins/workspace/build/lib/libcaffe2.so: undefined reference to `at::native::rnn_tanh_mkldnn_stub(at::Tensor const&, at::Tensor const&, c10::ArrayRef<at::Tensor>, bool, long, double, bool, bool, bool)'
Jan 15 01:11:14 /var/lib/jenkins/workspace/build/lib/libcaffe2.so: undefined reference to `at::native::lstm_mkldnn_stub(at::Tensor const&, c10::ArrayRef<at::Tensor>, c10::ArrayRef<at::Tensor>, bool, long, double, bool, bool, bool)'
Jan 15 01:11:14 /var/lib/jenkins/workspace/build/lib/libcaffe2.so: undefined reference to `at::native::gru_mkldnn_stub(at::Tensor const&, at::Tensor const&, c10::ArrayRef<at::Tensor>, bool, long, double, bool, bool, bool)'
Jan 15 01:11:14 /var/lib/jenkins/workspace/build/lib/libcaffe2.so: undefined reference to `at::native::rnn_relu_mkldnn_stub(at::Tensor const&, at::Tensor const&, c10::ArrayRef<at::Tensor>, bool, long, double, bool, bool, bool)'
Jan 15 01:11:14 collect2: error: ld returned 1 exit status
```

File: aten/src/ATen/native/mkldnn/RNN.cpp [hunk 1]
```diff
@@ -0,0 +1,757 @@
...
+    const Tensor& input_r, TensorList weight, int64_t weight_stride0,
+    const Tensor& hx, const Tensor& cx, const Tensor& output_r,
+    const Tensor& grad_output_r, const Tensor& grad_hy_r, const Tensor& grad_cy_r,
+    int64_t fn_mode, int64_t fn_hidden_size,
+    int64_t fn_num_layers, bool batch_first, double fn_dropout,
+    bool fn_train, bool fn_bidirectional, IntList fn_batch_sizes,
+    const Tensor& fn_dropout_state, TensorList reserve,
+    std::array<bool, 4> output_mask) {
+  AT_ERROR("mkldnn_rnn_backward: ATen not compiled with MKLDNN support");
+}
```
Comment (ezyang) [inline aten/src/ATen/native/mkldnn/RNN.cpp]: When mkldnn is not enabled, you still have to have dummy implementations of the stubs so linking works.

**Comment by mingfeima** [comment]
@ezyang thx for the head up!
Anyway, i still have some left jobs unfinished for this PR because of the complaint on inference performance. i will come back to this thread once inference shows reasonable speedup.

**Comment by cpuhrsch** [comment]
@mingfeima - If you're back, could I ask you to rebase this PR and @JianpingChen066 could I ask you to close https://github.com/pytorch/pytorch/pull/16491 in favor of a rebase of this PR? This can help us get rid of the duplication.

**Comment by cpuhrsch** [comment]
Can we add a test that covers these branches specifically and ensures they conform to some expected behavior?

**Comment by cpuhrsch** [comment]
In general, could we not use MKLDNN hooks? I know it is nice to change whether we want to use the library at runtime, but it adds a lot of complexity, for something we wouldn't want to turn off at runtime anyway (or maybe not?).

**Comment by mingfeima** [comment]
Recent change focused on `LSTM` inference performance speedup, which implements

- mkldnn rnn weight cache for inference
- mkldnn primitive cache
- mkldnn primitive descriptor cache
- Directly use mkldnn memory with native PyTorch tensor handle

The last two are specially for small time step sequence, e.g. `T=1`

With these changes, LSTM inference on `T=1` (the last scenario in the benchmark script) using single core improves `25%`, tested on Xeon 6148 (2*20 cores @2.5GHz), with this [script](https://github.com/mingfeima/pytorch-rnn-benchmark).
```bash
### inference mode

### using KMP_AFFINITY=granularity=fine,compact,1,0
### using KMP_BLOCKTIME=1
### using numactl --physcpubind=0-19 --membind=0


### using OMP_NUM_THREADS=20
mkldnn: [T,N,I,H] = [15,1,250,200], time: 0.67 (ms), SPS: 1482.55
cpu   : [T,N,I,H] = [15,1,250,200], time: 1.75 (ms), SPS: 570.93

### using OMP_NUM_THREADS=1
mkldnn: [T,N,I,H] = [15,1,250,200], time: 0.59 (ms), SPS: 1682.24
cpu   : [T,N,I,H] = [15,1,250,200], time: 1.86 (ms), SPS: 537.57

### using OMP_NUM_THREADS=1
### using layer=15
mkldnn: [T,N,I,H] = [1,1,250,200], time: 1.70 (ms), SPS: 586.57
cpu   : [T,N,I,H] = [1,1,250,200], time: 2.12 (ms), SPS: 470.79

### using OMP_NUM_THREADS=1
### using layer=1
mkldnn: [T,N,I,H] = [1,1,250,200], time: 0.15 (ms), SPS: 6531.25
cpu   : [T,N,I,H] = [1,1,250,200], time: 0.19 (ms), SPS: 5190.26
```
An additional `~30%` performance improvement is expected once MKLDNN is upgraded to `v0.19`, current version is `v0.17.3`

**Comment by cpuhrsch** [comment]
Great! Thank you @mingfeima.

Before we pull this back in for further performance testing, could I ask you to remove the hooks that enable the runtime MKL-DNN dispatch? They add a lot of complication and we run the risk of having users not use the more efficient library due.

**Comment by mingfeima** [comment]
@cpuhrsch mkldnn runtime hook is removed now.

**Comment by cpuhrsch** [comment]
@mingfeima: Test errors look real. `test_nn_LSTM` fails.

Just in case:

@pytorchbot rebase this please

**Comment by cpuhrsch** [comment]
(Accidentally closed the PR>..)

@pytorchbot rebase this please

File: torch/nn/modules/rnn.py [hunk 2]
```diff
@@ -200,10 +228,14 @@ def forward(self, input, hx=None):
             # the user believes he/she is passing in.
             hx = self.permute_hidden(hx, sorted_indices)
 
+        is_tranposed = self.use_transposed_weights and not self.training
+        self.transpose_paramters(is_tranposed)
+        _weights = self._transposed_weights if is_tranposed else self.get_flat_weights()
```
Comment (apaszke) [inline torch/nn/modules/rnn.py]: This is very much non-standard, and we should not hack such a functionality into every module individually without having a consistent library-wide consensus on how to implement support for different weight layouts (this includes convolutional modules in particular).

Comment (apaszke) [reply torch/nn/modules/rnn.py]: Apart from that, this will *unexpectedly* change the weight format without the user even asking for it, which is definitely not what any good library should do. This should be at least controlled by a flag which is off by default to keep backward compatibility, but we should really figure out a more widely applicable approach.

**Review by apaszke — CHANGES_REQUESTED** [review]
The weight format changes are unacceptable in the current form. It's BC-breaking and there's no way to turn it off.

File: torch/nn/modules/rnn.py (hunk 2 above)
Comment (mingfeima) [reply torch/nn/modules/rnn.py]: Yes, indeed this is not a decent design, it is mostly a temporary hack.

My complete idea to is isolate mkldnn storage from cpu storage, so we have an `MKLDNNTensorImpl`, kind of like `SparseTensorImpl`, and user may invoke mkldnn by `tensor.to_mkldnn()`, kind of like `tensor.to_sparse()`. This should be no hazard to both autograd and jit.

The mkldnn storage will basically handle two things:
1. mkldnn layout cache between forward and backward during training
2. mkldnn layout cache for inference

It will bring performance boost for `rnn` as well `convolution`, both training and inference.

Any comments on such design? @apaszke @cpuhrsch

Comment (jgong5) [reply torch/nn/modules/rnn.py]: #16038 documented the discussion on the idea.

**Comment by cpuhrsch** [comment]
@mingfeima - conflicts and test failures are real. Could I ask you to rebase and fix the issue within 'TestNN. test_RNN_dropout'?. Thank you.

File: torch/nn/modules/rnn.py (hunk 2 above)
Comment (cpuhrsch) [reply torch/nn/modules/rnn.py]: @mingfeima - this is a fairly complicated topic. Introducing a new tensor type isn't a path well defined for now. So I suggest we add a flag to prevent what @apaszke is pointing out. This is different from enabling/disabling MKL-DNN at runtime however. This could be a flag we pass as an argument and set to false by default. Let's worry about that once the test failures are resolved and we can confirm the performance gains internally.

**Comment by mingfeima** [comment]
@cpuhrsch i have fixed test case failure on `test_jit.py` `test_nn_LSTM`.
also i add a flag `is_transposed` which is set `False` by default, it will be turned on only when

- mkldnn is enabled
- the input can be handled by mkldnn
- inference mode (training mode does not require this)

so for other cases, e.g. using cuda device, the weights won't be transposed.

File: torch/nn/modules/rnn.py (hunk 2 above)
Comment (apaszke) [reply torch/nn/modules/rnn.py]: Adding a new tensor type is definitely an option. To be honest I would be quite unhappy even if we were to have a flag that controls this, because it sets an ugly precedent for other cases like this one. We should have a cleaner solution that covers more cases.

Comment (mingfeima) [reply torch/nn/modules/rnn.py]: Totally understand that! Upper level interface should be restricted by library dependent hacking.

You can now consider this PR as experimental to verify the `LSTM` performance on your production models. If the performance is not satisfactory, we will continue working on the mkldnn library part.

We are trying to add a new `TensorTypeId` in `c10` for mkldnn. Once it is done, nothing from `rnn.py` needs to be changed, mkldnn weight caching would be taken care of by the opaque data handle. Thus the mkldnn rnn function would be as clean as cudnn ones.

**Comment by bddppq** [comment]
@mingfeima @Jianhui-Li Per our offline meeting, we would like to move this to the mkldnn tensor path,  any chance you could work on it? :-)

**Comment by Jianhui-Li** [comment]
@bddppq Yes this PR needs to be refactored to based on the new MKL-DNN backend/tensor design.

**Comment by mingfeima** [comment]
@bddppq OOO last week due to family emergency. Will refactor this PR shortly.
I guess inference path will be enough?

**Comment by bddppq** [comment]
@mingfeima I hope everything is well with you and your family.
Thanks for working on it. Yeah for my use case inference will be sufficient.

**Review by apaszke — CHANGES_REQUESTED** [review]
Adding custom functions like `transpose_parameters` to `nn` modules is still a bit much. We need this for many more modules than CPU-side RNNs, so we should discuss a unified approach before we attempt to make changes to the standard library.

**Comment by mingfeima** [comment]
> Adding custom functions like `transpose_parameters` to `nn` modules is still a bit much. We need this for many more modules than CPU-side RNNs, so we should discuss a unified approach before we attempt to make changes to the standard library.

@apaszke Sure, i am rewriting the whole staff, `transpose_parameters` won't be needed and no touch to `torch/nn/modules/rnn.py`. The weight cache will be done via `torch/utils/mkldnn.py` in the same way which is proposed by @bddppq in `conv2d` weight cache. How does this sound?
```python
form torch import nn
from torch.utils import mkldnn as mkldnn_utils
rnn = nn.LSTM(10, 20)
# weight from mkldnn_rnn will be reordered and cached for inference
mkldnn_rnn = mkldnn_utils.to_mkldnn(rnn)
```

**Comment by apaszke** [comment]
Yeah, the out-of-line API like this is fine to have. My only problem is changing the core modules before we have a clear direction that can work in a more general setting. Thanks! :+1:

**Comment by mingfeima** [comment]
use new pr #26387
